src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,475 @@
+"""
+struct Row{T}
+    fun::T
+end
+
+normalize_selection(idx::AbstractIndex, sel) = idx[sel]
+
+function normalize_selection(idx::AbstractIndex, sel::ColumnIndex)
+    c = idx[sel]
+    return c => ColRename() => _names(idx)[c]
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex, Symbol})
+    c = idx[first(sel)]
+    return c => ColRename() => last(sel)
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:Any,<:Pair{<:Base.Callable,Symbol}})
+    c = first(sel)
+    return idx[c] => last(sel)
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:Any,<:Pair{<:Row,Symbol}})
+    c = first(sel)
+    return idx[c] => last(sel)
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex,<:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    newcol = Symbol(_names(idx)[c], "_", funname(fun))
+    return c => fun => newcol
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex, <:Row})
+    c = idx[first(sel)]
+    row = last(sel)
+    newcol = Symbol(_names(idx)[c], "_", funname(row.fun))
+    return c => row => newcol
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:Any, <:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    if length(c) > 3
+        newcol = Symbol(join(@views _names(idx)[c[1:2]], '_'), "_etc_", funname(fun))
+    else
+        newcol = Symbol(join(view(_names(idx), c), '_'), '_', funname(fun))
+    end
+    return c => fun => newcol
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:Any, <:Row})
+    c = idx[first(sel)]
+    row = last(sel)
+    if length(c) > 3
+        newcol = Symbol(join(@views _names(idx)[c[1:2]], '_'), "_etc_", funname(row.fun))
+    else
+        newcol = Symbol(join(@views _names(idx)[c], '_'), '_', funname(row.fun))
+    end
+    return c => row => newcol
+end
+
+function select_transform!(nc::Union{Pair{Int, Pair{ColRename, Symbol}},
+                                     Pair{<:Union{Int, AbstractVector{Int}},
+                                          <:Pair{<:Union{Base.Callable, Row}, Symbol}}},
+                           df::DataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Symbol, Any}, copycols::Bool)
+    col_idx = first(nc)
+    transform_spec = last(nc)
+    newname = last(transform_spec)
+    if !isnothing(transformed_cols[newname])
+        @assert !hasproperty(newdf, newname)
+    end
+    if nc isa Pair{Int, Pair{ColRename, Symbol}}
+        newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    elseif nc isa Pair{Int, <:Pair{<:Row, Symbol}}
+        newdf[!, newname] = (first(transform_spec).fun).(df[!, col_idx])
+    elseif nc isa Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+        res = first(transform_spec)(df[!, col_idx])
+        newdf[!, newname] = res isa AbstractVector ? res : [res]
+    elseif nc isa Pair{<:AbstractVector{Int}, <:Pair{<:Row, Symbol}}
```

Comment:
This is tempting, but we need a separate `if` to handle `0`-column selection in each-row version (as it has a special behavior). Also we need two separate behaviors for single column and multi column selections for each-row version (so there would be two methods for `Rows{F}` functor).

Given this I thought that it would be more readable (albeit less elegant) to have everything explicit in branches. In this way the code will be easier to digest for less experienced users (who might even not heard of functors).

Given this do you think it is worth to make this change?

Reply:
Yeah, I suspected it wasn't that simple. I'd find it more elegant -- two methods doesn't sound like the end of the world.

Reply:
Have you tried it in the end?

Reply:
OK - I have pushed how it can be implemented so that you can compare (we can revert this commit if needed).

I have also changed `Row` to `ByRow`.